### ISSUE TITLE
UserWarning for train dpo with lora: None of the inputs have requires_grad=True. Gradients will be None

### ISSUE THREAD

**Issue by xkw666** [huggingface/trl#2486]
### System Info

- Platform: Linux-6.5.0-41-generic-x86_64-with-glibc2.35
- Python version: 3.12.8
- PyTorch version: 2.2.2
- CUDA device(s): NVIDIA A100-PCIE-40GB, NVIDIA A100-PCIE-40GB
- Transformers version: 4.46.3
- Accelerate version: 1.2.1
- Datasets version: 3.2.0
- HF Hub version: 0.26.5
- TRL version: 0.13.0.dev0
- bitsandbytes version: 0.45.0
- DeepSpeed version: 0.16.1
- Diffusers version: not installed
- Liger-Kernel version: not installed
- LLM-Blender version: not installed
- OpenAI version: not installed
- PEFT version: 0.14.0

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder
- [ ] My own task or dataset (give details below)

### Reproduction

```python
import argparse,os

import torch
from datasets import load_dataset
from time import sleep
from transformers import AutoModelForCausalLM, AutoTokenizer

from trl import (
    DPOConfig,
    DPOTrainer,
    ModelConfig,
    ScriptArguments,
    TrlParser,
    get_kbit_device_map,
    get_peft_config,
    get_quantization_config,
)
from trl.trainer.utils import SIMPLE_CHAT_TEMPLATE


def main(script_args, training_args, model_args):
    ################
    # Model & Tokenizer
    ###################
    torch_dtype = (model_args.torch_dtype if model_args.torch_dtype in ["auto", None] else getattr(torch, model_args.torch_dtype))
    quantization_config = get_quantization_config(model_args)

    model_kwargs = dict(
        revision=model_args.model_revision,
        attn_implementation=model_args.attn_implementation,
        torch_dtype=torch_dtype,
        use_cache=False if training_args.gradient_checkpointing else True,
        device_map=get_kbit_device_map() if quantization_config is not None else 'cuda',
        quantization_config=quantization_config,)
    
    model = AutoModelForCausalLM.from_pretrained(
        model_args.model_name_or_path, trust_remote_code=model_args.trust_remote_code, **model_kwargs
    )
    if hasattr(model, "enable_input_require_grads"):
        print('enable_input_require_grads')
        model.enable_input_require_grads() ## To avoid error https://github.com/huggingface/trl/issues/731

    peft_config = get_peft_config(model_args)
    if peft_config is None:
        ref_model = AutoModelForCausalLM.from_pretrained(
            model_args.model_name_or_path, trust_remote_code=model_args.trust_remote_code, **model_kwargs
        )
    else:
        ref_model = None
        
    tokenizer = AutoTokenizer.from_pretrained(
        model_args.model_name_or_path, trust_remote_code=model_args.trust_remote_code
    )

    if tokenizer.pad_token is None:
        tokenizer.pad_token = tokenizer.eos_token
    if tokenizer.chat_template is None:
        tokenizer.chat_template = SIMPLE_CHAT_TEMPLATE
    if script_args.ignore_bias_buffers:
        # torch distributed hack
        model._ddp_params_and_buffers_to_ignore = [
            name for name, buffer in model.named_buffers() if buffer.dtype == torch.bool
        ]

    ################
    # Dataset
    ################
    dataset = load_dataset(script_args.dataset_name, name=script_args.dataset_config)
    print(dataset)
    ##########
    # Training
    ################
    print( training_args.eval_strategy)
    trainer = DPOTrainer(
        model,
        ref_model,
        args=training_args,
        train_dataset=dataset[script_args.dataset_train_split],
        eval_dataset=dataset[script_args.dataset_test_split] if training_args.eval_strategy != "no" else None,
        processing_class=tokenizer,
        peft_config=peft_config,
    )

    trainer.train()

    if training_args.eval_strategy != "no":
        metrics = trainer.evaluate()
        trainer.log_metrics("eval", metrics)
        trainer.save_metrics("eval", metrics)

    # Save and push to hub
    trainer.save_model(training_args.output_dir)
    output_dir = os.path.join(script_args.output_dir, "final_checkpoint")
    trainer.model.save_pretrained(output_dir)
    # if training_args.push_to_hub:
    #     trainer.push_to_hub(dataset_name=script_args.dataset_name)


def make_parser(subparsers: argparse._SubParsersAction = None):
    dataclass_types = (ScriptArguments, DPOConfig, ModelConfig)
    if subparsers is not None:
        print("=="*50)
        parser = subparsers.add_parser("dpo", help="Run the DPO training script", dataclass_types=dataclass_types)
    else:
        parser = TrlParser(dataclass_types)
    return parser


if __name__ == "__main__":
    parser = make_parser()
    script_args, training_args, model_args = parser.parse_args_and_config()
    print('\n',model_args)
    print('\n',training_args)
    print('\n',script_args)
    main(script_args, training_args, model_args)


```

outputs:

```
/anaconda3/envs/Safe/lib/python3.12/site-packages/torch/utils/checkpoint.py:90: UserWarning: None of the inputs have requires_grad=True. Gradients will be None
    ...
```

### Expected behavior

I used the official example to train dpo with lora, and encountered `UserWarning: None of the inputs have requires_grad=True. Gradients will be None.` I found similar problems in issus and I added this step  `model.enable_input_require_grads()` in the code, but this warning still appears. **I wonder if this will affect the training results?**

### Checklist

- [X] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [X] I have included my system information
- [X] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [X] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [X] Any traceback provided is complete

**Comment by AIR-hl** [comment]
According to my experience, there is a problem with your data, and None data exist in it.

**Comment by Ciao-CA** [comment]
I have the same problem

**Comment by karlcuinju** [comment]
Any solution now？

**Comment by ttyyc** [comment]
> ### System Info
> * Platform: Linux-6.5.0-41-generic-x86_64-with-glibc2.35
> * Python version: 3.12.8
> * PyTorch version: 2.2.2
> * CUDA device(s): NVIDIA A100-PCIE-40GB, NVIDIA A100-PCIE-40GB
> * Transformers version: 4.46.3
> * Accelerate version: 1.2.1
> * Datasets version: 3.2.0
> * HF Hub version: 0.26.5
> * TRL version: 0.13.0.dev0
> * bitsandbytes version: 0.45.0
> * DeepSpeed version: 0.16.1
> * Diffusers version: not installed
> * Liger-Kernel version: not installed
> * LLM-Blender version: not installed
> * OpenAI version: not installed
> * PEFT version: 0.14.0
> 
> ### Information
> * [x]  The official example scripts[ ]  My own modified scripts
> 
> ### Tasks
> * [x]  An officially supported task in the `examples` folder[ ]  My own task or dataset (give details below)
> 
> ### Reproduction
> import argparse,os
> 
> import torch
> from datasets import load_dataset
> from time import sleep
> from transformers import AutoModelForCausalLM, AutoTokenizer
> 
> from trl import (
>     DPOConfig,
>     DPOTrainer,
>     ModelConfig,
>     ScriptArguments,
>     TrlParser,
>     get_kbit_device_map,
>     get_peft_config,
>     get_quantization_config,
> )
> from trl.trainer.utils import SIMPLE_CHAT_TEMPLATE
> 
> 
> def main(script_args, training_args, model_args):
>     ################
>     # Model & Tokenizer
>     ###################
>     torch_dtype = (model_args.torch_dtype if model_args.torch_dtype in ["auto", None] else getattr(torch, model_args.torch_dtype))
>     quantization_config = get_quantization_config(model_args)
> 
>     model_kwargs = dict(
>         revision=model_args.model_revision,
>         attn_implementation=model_args.attn_implementation,
>         torch_dtype=torch_dtype,
>         use_cache=False if training_args.gradient_checkpointing else True,
>         device_map=get_kbit_device_map() if quantization_config is not None else 'cuda',
>         quantization_config=quantization_config,)
>     
>     model = AutoModelForCausalLM.from_pretrained(
>         model_args.model_name_or_path, trust_remote_code=model_args.trust_remote_code, **model_kwargs
>     )
>     if hasattr(model, "enable_input_require_grads"):
>         print('enable_input_require_grads')
>         model.enable_input_require_grads() ## To avoid error https://github.com/huggingface/trl/issues/731
> 
>     peft_config = get_peft_config(model_args)
>     if peft_config is None:
>         ref_model = AutoModelForCausalLM.from_pretrained(
>             model_args.model_name_or_path, trust_remote_code=model_args.trust_remote_code, **model_kwargs
>         )
>     else:
>         ref_model = None
>         
>     tokenizer = AutoTokenizer.from_pretrained(
>         model_args.model_name_or_path, trust_remote_code=model_args.trust_remote_code
>     )
> 
>     if tokenizer.pad_token is None:
>         tokenizer.pad_token = tokenizer.eos_token
>     if tokenizer.chat_template is None:
>         tokenizer.chat_template = SIMPLE_CHAT_TEMPLATE
>     if script_args.ignore_bias_buffers:
>         # torch distributed hack
>         model._ddp_params_and_buffers_to_ignore = [
>             name for name, buffer in model.named_buffers() if buffer.dtype == torch.bool
>         ]
> 
>     ################
>     # Dataset
>     ################
>     dataset = load_dataset(script_args.dataset_name, name=script_args.dataset_config)
>     print(dataset)
>     ##########
>     # Training
>     ################
>     print( training_args.eval_strategy)
>     trainer = DPOTrainer(
>         model,
>         ref_model,
>         args=training_args,
>         train_dataset=dataset[script_args.dataset_train_split],
>         eval_dataset=dataset[script_args.dataset_test_split] if training_args.eval_strategy != "no" else None,
>         processing_class=tokenizer,
>         peft_config=peft_config,
>     )
> 
>     trainer.train()
> 
>     if training_args.eval_strategy != "no":
>         metrics = trainer.evaluate()
>         trainer.log_metrics("eval", metrics)
>         trainer.save_metrics("eval", metrics)
> 
>     # Save and push to hub
>     trainer.save_model(training_args.output_dir)
>     output_dir = os.path.join(script_args.output_dir, "final_checkpoint")
>     trainer.model.save_pretrained(output_dir)
>     # if training_args.push_to_hub:
>     #     trainer.push_to_hub(dataset_name=script_args.dataset_name)
> 
> 
> def make_parser(subparsers: argparse._SubParsersAction = None):
>     dataclass_types = (ScriptArguments, DPOConfig, ModelConfig)
>     if subparsers is not None:
>         print("=="*50)
>         parser = subparsers.add_parser("dpo", help="Run the DPO training script", dataclass_types=dataclass_types)
>     else:
>         parser = TrlParser(dataclass_types)
>     return parser
> 
> 
> if __name__ == "__main__":
>     parser = make_parser()
>     script_args, training_args, model_args = parser.parse_args_and_config()
>     print('\n',model_args)
>     print('\n',training_args)
>     print('\n',script_args)
>     main(script_args, training_args, model_args)
> outputs:
> 
> ```
> /anaconda3/envs/Safe/lib/python3.12/site-packages/torch/utils/checkpoint.py:90: UserWarning: None of the inputs have requires_grad=True. Gradients will be None
>     ...
> ```
> 
> ### Expected behavior
> I used the official example to train dpo with lora, and encountered `UserWarning: None of the inputs have requires_grad=True. Gradients will be None.` I found similar problems in issus and I added this step `model.enable_input_require_grads()` in the code, but this warning still appears. **I wonder if this will affect the training results?**
> 
> ### Checklist
> * [x]  I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))[x]  I have included my system information[x]  Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))[x]  Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))[x]  Any traceback provided is complete

No need to worry about affecting the training outcome, as the visual part of the model is frozen, and the input gradient for this part is naturally 0.

**Comment by albertvillanova** [comment]
Hey @xkw666, sorry for the slow reply — we're catching up on the backlog. 

That warning is harmless and does not affect your DPO+LoRA training: it comes from the *reference* forward pass, which TRL runs with the LoRA adapter disabled under `torch.no_grad()` (https://github.com/huggingface/trl/blob/44596766101a123f1a3ff70b0aefa54d926179ad/trl/trainer/dpo_trainer.py#L1328-L1331). Since that pass needs no gradients, PyTorch's checkpoint helper prints "None of the inputs have requires_grad=True", but your policy adapters still receive gradients and train normally. Your manual `model.enable_input_require_grads()` is also redundant here — DPOTrainer already calls it internally for PEFT + gradient checkpointing (https://github.com/huggingface/trl/blob/44596766101a123f1a3ff70b0aefa54d926179ad/trl/trainer/dpo_trainer.py#L673-L674). On current main the warning is additionally suppressed for offline DPO (we disable checkpointing during the reference pass and default to `use_reentrant=False`), so an upgrade should make it disappear entirely. 

Closing as not-a-bug — happy to re-open if you ever see the loss actually failing to move.
